WIP: Refactor of subsampling module

dotsdl opened the pull request:
Addresses #95, #79, #91 

Required for merge:
- [x] reference documentation for the functions included in the `alchemlyb.preprocessing.subsampling` module
- [ ] example notebook showing usage, as well as rationale for usage, of each function
- [x] tests across more datasets in `alchemtest`
- [ ] changelog entry

This is an API breaking change.

All three of our subsampling functions should now be able to consume any DataFrame in our two standard `u_nk` and `dHdl` forms and produce meaningful results. Each of these functions now performs a `groupby` with all non-time indexes, and separately performs its operation on each group. The resulting subsampled groups are concatenated and returned.

Some notes on the implementation:
1. The `series` kwarg is now the `column` kwarg for  `statistical_inefficiency`, `equilibrium_detection`. This can take either a column name in the DataFrame or a Series object (the index of the Series must match *exactly* that of the DataFrame). Taking a Series object allows for code that uses the existing usage of e.g. `statistical_inefficiency(u_nk, u_nk[<column_name>])` to still work as before.
2. `statistical_inefficiency` and `equilibrium_detection` now have a `how` kwarg. I expect this to be the main usage we suggest for users of these functions. I took a look at how `alchemical_analysis` [currently handles `dHdl` and `u_nk` cases](https://github.com/MobleyLab/alchemical-analysis/blob/master/alchemical_analysis/alchemical_analysis.py#L159), and to replicate that as closely as possible, we recommend using `how='right'` for `u_nk` data, and `how='sum'` for `dHdl` data. Please see the docstrings for `statistical_inefficiency`, `equilibrium_detection` for details on these treatments.
3. Either `column` or `how` must be used for `statistical_inefficiency`, `equilibrium_detection`. The function can no longer default to simple slicing if only `data` is given as an input. This is to avoid (potentially new) users from applying the functions without specifying these inputs with no obvious failure, but with the function doing nothing to decorrelate their data (essentially a silent failure).
4. I've added the `force` kwarg to `statistical_inefficiency`, `equilibrium_detection`. The only exception these functions will skip with `force=True` is one due to duplicate time values in the groupings, as this tends to indicate two or more timeseries have been dumped together, which is unlikely what one wants to do with these subsampling functions that assume correlation to work as intended.
5. Setting `return_calculated=True` for `statistical_inefficiency`, `equilibrium_detection` will return calculated values used for subsampling each lambda grouping:
    - `statistical_inefficiency`: `statinef`
    - `equilibrium_detection`: `t`, `statinef`, `Neff_max`

Some other nice things:
1. I will be adding the minimal assumptions the standard forms make to their doc page as part of this change's PR. One of the assumptions I want to propagate is that the outermost index is *always* time, or at least something implying time (frame, sample, whatever). It needn't be called 'time', however, for all the tooling to work.
2. These subsamplers now explicitly call out that they *can't be expected* to preserve any custom ordering of rows. They will always be doing sorting of some kind, so the time index values are meaningful.

dotsdl committed:
Refactor of subsampling module

All three of our subsampling functions should now be able to consume any
dataframe in our two standard u_nk and dHdl forms and produce meaningful
results. Each of these functions now performs a groupby with all
non-time indexes, and separately performs its operation on each group.

Some notes on the implementation:
1. The `series` kwarg is now the `column` arg for
   `statistical_inefficiency`, `equilibrium_detection`. A column name is
   now what we use, as requiring a column of the dataframe offers
   guarantees around the index matching, ordering, etc. that we had to
   do backflips to guarantee before. It is still possible to use another
   series as the basis for either operation, but it should be grafted on
   to the dataframe as a column to do so.
2. Since `column` is an arg, it is now required for the above functions.
   Therefore, those functions no longer behave like `slicing` when
   `column` isn't specified. My logic with breaking this is to favor
   explicitness and to avoid surprising users with behavior that throws
   no warnings but silently may not be doing what they want.
3. I've added some kwargs to `force` ignoring exceptions. The only
   exception these functions throw now is one due to duplicate time values
   in the groupings, as this tends to indicate two or more timeseries have
   been dumped together, which is unlikely what one wants to do with these
   subsampling functions that assume correlation to work as intended.

Some other nice things:
1. I will be adding the minimal assumptions the standard forms make to
their doc page as part of this change's PR. One of the assumptions I
want to propagate is that the outermost index is *always* time, or at
least something implying time (frame, sample, whatever). It needn't be
called 'time', however, for all the tooling to work.
2. These subsamplers now explicitly call out that they *can't be
expected* to preserve any custom ordering of rows. They will always be
doing sorting of some kind, so the time index values are meaningful.

dotsdl commented:
This [gist](https://gist.github.com/dotsdl/a526a5827781f6e9c05f229e81487b06) demonstrates the behavior of the refactored functions.

dotsdl committed:
Added tests for subsampling on most of our datasets

Not getting all passes yet, and not sure I'm happy with how these work.
There are clear shortcomings with the static column approach, requiring
perhaps manual effort on the part of the user to work around.
This is counter to alchemlyb's philosophy of making the 'most right'
thing easy to do.

dotsdl committed:
Docs for subsamplers reflect a change in direction

In the spirit of wanting to make it easy for users to do the 'most
right' thing, I have looked at the details of what `alchemical-analysis`
currently does for subsampling with statisticalInefficiency, and it uses
an approach similar to the one documented in each of our
`statistical_inefficiency` and `equilbrium_detection` functions now.

Next steps are to *actually* implement this functionality.
We are going to make our lives easier by adding an `alchemform`
attribute to dataframes our parsers produce, and add this to the
standard form spec. This attribute will have 'dHdl' or 'u_nk' as its
value, depending on the form of data we're dealing with.
Our estimators can also use this as a check or warning.

dotsdl committed:
statistical_inefficiency should now behave according to its doc

`how` methods implemented; these require explicit tests yet.
We should also add logging to each of these functions, using a module-level logger.
This would allow for auditability later in case downstream issues are
discovered.

dotsdl committed:
Tests passing for statistical_inefficiency

About to break out the generic bits to make the same changes to
equilibrium_detection.

dotsdl committed:
equilibrium_detection now shares same basic code as statistical_inefficiency

All tests pass, but we should try and think carefully about tests that
we really want to add to ensure these are working as expected.

dotsdl committed:
alchemform hits a stumbling block

Using something like `pd.concat` on a set of DataFrames doesn't
propagate forward the `.attrs` dictionary, so having an `alchemform` key
here created by the parsers is not sufficient for signalling to
preprocessors the type of data.

We will likely remove the 'auto' option from the subsamplers and insist
on explicit values for `how`. This is not *necessarily* bad, as it
promotes some understanding for the user of what they are choosing to do
here.

dotsdl committed:
Removed use of `alchemform`; users must explicitly choose their `how`

Alternatively, they can choose a `column`.
I think this is appropriate, as it is still important that users are
consciously aware of some of the high-level choices the software is making.
There will always be touchpoints like this, though we should try to
minimize them where possible.

dotsdl committed:
Added some test coverage for return_calculated, using a Series

These changes apply to tests of `statistical_inefficiency`, `equilibrium_detection`.
I also set `nskip=5` for `equilibrium_detection` cases, which required
some changes to test parameters.
This change speeds up the tests, however, at no real coverage cost.

dotsdl commented:
The changes I've made to the codebase are ready for review! I've updated the PR description with the current state of the design, as it took a few iterations for me to arrive at. I would like some feedback on this design as soon as possible so changes can be made to accommodate needs. This PR is also a bit of a blocker for #94 and #99, since subsampling is a key component of the `alchemlyb` workflow.

I am now working on an example notebook that will make its way into our examples docs; this will showcase how to use these functions, as well as when each makes sense to use. I don't consider this notebook a requirement for review of the implementation, but consider it important to make sure the feel of this approach is right. I think I can turn it around and add to this PR by the end of this week.

dotsdl commented:
@orbeckst and @mrshirts I've requested your eyes to give this implementation a pass. In particular, I would like a check on the `how` parameter options for `statistical_inefficiency` and `equilibrium_detection`, and whether they make sense or appear correct.

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -1,26 +1,79 @@
+    candidates = set(group.columns) - (set(tried) | {name})
+
+    if not candidates:
+        raise CorrelationError("No column in the dataset could be used"
```
This could be `RuntimeError`.

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -34,63 +87,198 @@ def slicing(df, lower=None, upper=None, step=None, force=False):
+    correlations within each group of samples. The options are as follows:
+
+        'right'
+            The recommended setting for 'u_nk' datasets; the column immediately
```
Who's doing the recommending and on what grounds?

If my student used alchemlyb and said "I used the recommended settings" then I will ask "recommendation based on what?" and then the student won't know what to say.

I'd like a more scholarly approach to recommendations. Or don't recommend anything.

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -34,63 +87,198 @@ def slicing(df, lower=None, upper=None, step=None, force=False):
+            is used. If there is no column to the left, then the column to the
+            right is used.  If there is no column corresponding to the group's
+            lambda index value, then 'random' is used for that group (see below).
+        'random'
```
Why this option?? This seems counterproductive to reproducible research. Does the user even know which column worked in the end? I'd much rather fail.

orbeckst replied:
I really do not like this option in a library that tries to be clean and basic. Let the users write their workflows and heuristics. 👎

dotsdl replied:
I agree that we would like to avoid randomness where possible. We could remove the `how='random'` parameter choice here, but we may still need to internally address the (admittedly rare?) corner case of a `u_nk` column not existing for a lambda state actually sampled. For the "right"/"left" approaches, if a corresponding column doesn't exist for the lambda state in question, the concept of right and left becomes more guesswork.

Some alternative approaches I can think of:
1. Find the "closest" column in terms of euclidean distance in lambdas, and then go to the right or left of that, but this logic seems to be pretty tortured to me, and I'm not sure if it too can fail in weird ways.
2. Choose the leftmost or rightmost column in the whole DataFrame, and march through columns in order if we hit failures to decorrelate.
3. Try to use the last column we subsampled on (previous iteration of [this loop](https://github.com/alchemistry/alchemlyb/pull/98/files/dc1b3ff106eb1d66f1b13555c6d1abe3d65c0e8d#diff-29b9db00345fef71312e85c1392b6fbeR157)).
4. Fail loudly, and declare this corner case not one we wish to address. The user could get around this by throwing out rows corresponding to lambda values there are no corresponding columns for.

dotsdl replied:
One way we can address some of the harder-to-pin-down choices made in alchemlyb is to use a logger (#34). I recognize that this doesn't excuse *strange* choices on the library's part, but if we have a logger in place then at least what the library did in a particular instance is auditable later.

orbeckst replied:
Logging would be a good step towards more transparency.

Maybe put in a commented-out logger call with a hint towards #34?

dotsdl replied:
Will do! I consider this an important component of the library.

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -34,63 +87,198 @@ def slicing(df, lower=None, upper=None, step=None, force=False):
+        for each group.
+    force : bool
+        Ignore checks that DataFrame is in proper form for expected behavior.
+    random_state : int
```
👎

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -34,63 +87,198 @@ def slicing(df, lower=None, upper=None, step=None, force=False):
-    no automatic conversion from a time to an index.
+    Raises
+    ------
+    CorrelationError
```
Not sure if custom error is justified.

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -102,101 +290,208 @@ def statistical_inefficiency(df, series=None, lower=None, upper=None, step=None,
 
+    References
+    ----------
+    [1] J. D. Chodera, W. C. Swope, J. W. Pitera, C. Seok, and K. A. Dill. Use of the weighted
```
I would use reST citations and have one central citations page. At the end of the day, I think users are more likely to read the online docs for this kind of information than when they read the raw doc strings.

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -102,101 +290,208 @@ def statistical_inefficiency(df, series=None, lower=None, upper=None, step=None,
-       ``pymbar.timeseries.statisticalInefficiency()``; previously, the statistical
-       inefficiency was _rounded_ (instead of ``ceil()``) and thus one could
-       end up with correlated data.
+       ``pymbar.timeseries.subsampleCorrelatedData()``; previously, the
```
Did pymbar change the function name or was this a mistake?

dotsdl replied:
This was a mistake in the docstring that I fixed here; the same mistake was not made in the equivalent line for `equilibrium_detection`.

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -102,101 +290,208 @@ def statistical_inefficiency(df, series=None, lower=None, upper=None, step=None,
+
+        return indices, calculated
+
+    return _decorrelator(
```
I like the more modular approach here.

dotsdl replied:
Thank you!

orbeckst commented on src/alchemlyb/preprocessing/subsampling.py:
```diff
@@ -102,101 +290,208 @@ def statistical_inefficiency(df, series=None, lower=None, upper=None, step=None,
+            is used. If there is no column to the left, then the column to the
+            right is used.  If there is no column corresponding to the group's
+            lambda index value, then 'random' is used for that group (see below).
+        'random'
```
Not a fan... 👎

orbeckst commented on src/alchemlyb/tests/test_preprocessing.py:
```diff
@@ -10,117 +10,284 @@
+        # results can vary slightly with different machines
+        # so possibly do
+        # delta = 10
+        # assert size - delta < len(sliced) < size + delta
```
You could also turn it into a floating point comparison, e.g., `size/len(sliced)` as the fraction and then just use `assert_almost_equal(...., decimals=2)` or whatever. Either way, if you already know that there can be variation then put it in the test and comment it.

orbeckst commented on src/alchemlyb/tests/test_preprocessing.py:
```diff
@@ -10,117 +10,284 @@
-    def test_conservative(self, data, size, conservative):
-        sliced = self.slicer(data, series=data.iloc[:, 0], conservative=conservative)
+    def test_conservative(self, data, size, conservative, how):
+        sliced = self.subsampler(data, how=how, conservative=conservative)
```
less stringent test?

orbeckst reviewed (commented):
Hi @dotsdl , I had a quick look and overall this looks like great work. Some comments inline.

dotsdl commented on docs/preprocessing/alchemlyb.preprocessing.subsampling.rst:
```diff
@@ -1,11 +1,268 @@
+
+    g \equiv (1 + 2\tau) > 1
+
+where :math:`\tau` is the autocorrelation time of the timeseries.
```
Ah, good call. Tau in this usage is measured in frames, or ordered rows, of a timeseries. I will change the wording to reflect this.

dotsdl commented on docs/preprocessing/alchemlyb.preprocessing.subsampling.rst:
```diff
@@ -1,11 +1,268 @@
+
+The ``how`` parameter indicates the choice of observable for performing the calculation of :math:`g`.
+
+* For :math:`u_{nk}` datasets, the choice of ``'right'`` is recommended: the column immediately to the right of the column corresponding to the group's lambda index value is used as the observable.
```
I'd be happy to, but they come from my attempt to emulate what is already done in [alchemical-analysis.py](https://github.com/MobleyLab/alchemical-analysis/blob/master/alchemical_analysis/alchemical_analysis.py#L159), and I don't know if a reference exists for the choices made there. @mrshirts may be able to provide some direction on this.

I do believe I understand the reason for the choices, however: in principle, any column in a timeseries would be fine to base decorrelation on, so long as the column is not uniformly zero or insensitive to time/frame for the given lambda window being sampled. 

1. The approach of choosing the column to the right of the lambda window being sampled for `u_nk` avoids cases where the `u_nk` of calculated lambdas are relative to the sampled one, giving all zeros for the column corresponding to the lambda actually sampled. An example of this observed:

```python
from alchemtest.gmx import load_water_particle_without_energy
from alchemlyb.parsing.gmx import extract_u_nk
import pandas as pd

ds = load_water_particle_without_energy()

print(pd.concat(extract_u_nk(i, T=300) for i in ds.data['AllStates']).sort_index(level=[1, 2, 0]))
```
This gives:
```
                               (0.0, 0.0)  ...    (1.0, 1.0)
time   coul-lambda vdw-lambda              ...              
0.0    0.0         0.0           0.000000  ...    148.347412
18.6   0.0         0.0           0.000000  ...   1983.809792
37.2   0.0         0.0           0.000000  ...    205.930556
55.8   0.0         0.0           0.000000  ...   4512.816690
74.4   0.0         0.0           0.000000  ...  25939.040202
...                                   ...  ...           ...
9913.8 1.0         1.0          45.353152  ...      0.000000
9932.4 1.0         1.0          41.207448  ...      0.000000
9951.0 1.0         1.0          44.375182  ...      0.000000
9969.6 1.0         1.0          45.505032  ...      0.000000
9988.2 1.0         1.0          37.779798  ...      0.000000

[20444 rows x 38 columns]


```
Note that if we chose any old column, for one of the groupings of the `('coul-lambda', 'vdw-lambda`)` indices, we would end up trying to decorrelate on all zeros. the `how="right"` choice sidesteps this (as does `how="left"`).

2. The approach of summing all `dHdl` columns is not as clear to me, and I wouldn't be surprised if I've misinterpreted the implementation in `alchemical-analysis.py`.

dotsdl commented on docs/preprocessing/alchemlyb.preprocessing.subsampling.rst:
```diff
@@ -1,11 +1,268 @@
+* For :math:`u_{nk}` datasets, the choice of ``'right'`` is recommended: the column immediately to the right of the column corresponding to the group's lambda index value is used as the observable.
+* For :math:`\frac{dH}{d\lambda}` datasets, the choice of ``'sum'`` is recommended: the columns are simply summed, and the resulting :py:class:`pandas.Series` is used as the observable.
+
+See the API documentation below on the possible values for ``how``, as well as more detailed explanations for each choice.
```
Agreed, I would like some assistance with this. See my notes above.

dotsdl commented on docs/preprocessing/alchemlyb.preprocessing.subsampling.rst:
```diff
@@ -1,11 +1,268 @@
+
+Equilibrium Detection
+---------------------
+The :func:`~alchemlyb.preprocessing.subsampling.equilibrium_detection` function subsamples each timeseries in the dataset using the equilibration detection approach developed by John Chodera (see reference below).
```
Understood, I will change this accordingly.

dotsdl commented:
Thank you for your review @orbeckst! Your comments are incredibly helpful in steering this refactor to a satisfying state; I will be addressing the points you have raised over the next several days.

orbeckst commented:
@dotsdl what do we need here to get this in? I’m asking because if we want to do breaking changes then I’d rather do it sooner than later. 

Or we do 0.4 #113 and then do this one for a 0.5 with an eye towards 1.0. Overall it looks to me that alchemlyb has sufficiently stabilized to just declare the API stable soon.

xiki-tempula commented:
I don't think there are significant features that are not implemented in the 2.0.0. But I do feel that the documentation could benefit.
